onnxruntime/core/session/utils.cc

```diff
@@ -348,12 +352,17 @@ Status CreateIExecutionProviderFactoryForEpDevices(const Environment& env,
                            "All OrtEpDevice values in ep_devices must have the same execution provider.");
   }
 
-  EpFactoryInternal* internal_factory = nullptr;
+  std::unique_ptr<IExecutionProviderFactory> provider_factory = nullptr;
+
   for (const OrtEpDevice* ep_device : ep_devices) {
-    // we expect the internal factory to be available for internal and provider bridge EPs, which is all we support.
-    internal_factory = env.GetEpFactoryInternal(ep_device->ep_factory);
-    if (!internal_factory) {
-      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "EP is not currently supported by this API");
+    if (provider_factory == nullptr) {
+      EpFactoryInternal* internal_factory = env.GetEpFactoryInternal(ep_device->ep_factory);
+
+      if (internal_factory) {
+        provider_factory = std::make_unique<InternalExecutionProviderFactory>(*internal_factory, ep_devices);
+      } else {
+        provider_factory = std::make_unique<PluginExecutionProviderFactory>(*ep_device->ep_factory, ep_devices);
```

Comment:
Should this break out of the `for` loop given we're providing `ep_devices` to the ctor and not just the current value?

Reply:
Breaking out here would prevent us from processing all OrtEpDevices (adding their ep_options to SessionOptions).

This statement is creating the factory in the first iteration of the loop only, so I believe it should be alright. That said, I do find it a bit confusing so I'll refactor.

Reply:
I refactored this create the factory outside of the loop. I think it makes the logic more understandable, but let me know what you think.